synth_node_bin/src/action/advanced_sn_for_s001.rs

```diff
@@ -0,0 +1,181 @@
+
+use anyhow::Result;
+use tokio::time::{interval, sleep, Duration};
+use ziggurat_zcash::{
+    protocol::{
+        message::Message,
+        payload::{addr::NetworkAddr, Addr, Nonce},
+    },
+    tools::{
+        message_filter::{Filter, MessageFilter},
+        synthetic_node::SyntheticNode,
+    },
+};
+
+use super::{ActionCfg, SynthNodeAction};
+
+// Configurable status printout interval.
+const DBG_INFO_LOG_INTERVAL_SEC: Duration = Duration::from_secs(10);
+const BROADCAST_INTERVAL_SEC: Duration = Duration::from_secs(120);
+
+pub(super) struct Action;
+
+pub(super) fn action() -> Box<dyn SynthNodeAction> {
+    Box::new(Action {})
+}
+
+#[async_trait::async_trait]
+impl SynthNodeAction for Action {
+    fn info(&self) -> &str {
+        "an advanced SN node created for RT-S001 which can do the following
+           - listens for inbound connections and prints the connection status periodically
+           - periodically requests and sends Ping/GetAddr/Addr messages to all peers"
+    }
+
+    fn config(&self) -> ActionCfg {
+        ActionCfg {
+            msg_filter: MessageFilter::with_all_auto_reply().with_getaddr_filter(Filter::Disabled),
+        }
+    }
+
+    #[allow(unused_variables)]
+    async fn run(&self, synth_node: &mut SyntheticNode, addr: SocketAddr) -> Result<()> {
+        // Sleep for 3 seconds before taking any actions.
+        sleep(Duration::from_millis(3000)).await;
+
+        let msg = Message::GetAddr;
+        tracing::info!("unicast {msg:?}\n");
+        if synth_node.unicast(addr, msg.clone()).is_err() {
+            tracing::warn!("failed to send {msg:?}\n");
+            anyhow::bail!("connection closed");
+        }
+
+        let mut dbg_info_interval = interval(DBG_INFO_LOG_INTERVAL_SEC);
+        let mut broadcast_msgs_interval = interval(BROADCAST_INTERVAL_SEC);
+
+        loop {
+            tokio::select! {
+                _ = dbg_info_interval.tick() => {
+                    trace_debug_info(synth_node).await;
+                },
+                _ = broadcast_msgs_interval.tick() => {
+                    let _ = broadcast_periodic_msgs(synth_node);
+                },
+                Ok((src, msg)) = synth_node.try_recv_message() => {
+                    handle_rx_msg(synth_node, src, msg).await?;
+                },
+            }
+        }
+    }
+}
+
+async fn trace_debug_info(synth_node: &SyntheticNode) {
+    let peer_infos = synth_node.connected_peer_infos();
+    let peer_cnt = peer_infos.len();
+
+    let mut log = format!("\nNumber of peers: {peer_cnt}\n");
+
+    // Let's sort by the connection's time value.
+    let mut peer_infos: Vec<_> = peer_infos.iter().collect();
+    peer_infos.sort_by(|a, b| a.1.stats().created().cmp(&b.1.stats().created()));
+
+    for (addr, info) in peer_infos.iter() {
+        let stats = info.stats();
+
+        // Align all possible IP addresses (both v4 and v6) vertically
+        const MAX_IPV6_ADDR_LEN: usize = 37;
```

Comment:
Why 37? 
8 groups of 4 digits with 7 `:` between them can give 39.

There are possible some other cases like [IPv4-mapped IPv6 address](https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5.2) where max length would be 45. That's why INET6_ADDRSTRLEN constant in Linux has 46 bytes (45 + 1 null terminating char).

Reply:
> 8 groups of 4 digits with 7 : between them can give 39.

ops, my miscalculation here haha

> There are possible some other cases like [IPv4-mapped IPv6 address](https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5.2) where max length would be 45. That's why INET6_ADDRSTRLEN constant in Linux has 46 bytes (45 + 1 null terminating char).

TIL about this - okay, then let's use 46 here

Reply:
Done